test/boost/big_decimal_test.cc

```diff
@@ -203,3 +203,116 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    test_cmp(big_decimal(max_scale, -1), big_decimal("-1"), std::strong_ordering::greater);
+    test_cmp(big_decimal(min_scale, -1), big_decimal("-1"), std::strong_ordering::less);
+
+    // Test the edges of the compare scale-compare algorithm.
+    // Note, scale has implicit inverted sign.
+    test_cmp(big_decimal(max_scale, huge_unscaled_number), big_decimal(min_scale, huge_unscaled_number), std::strong_ordering::less);
+    test_cmp(big_decimal(min_scale, huge_unscaled_number), big_decimal(max_scale, huge_unscaled_number), std::strong_ordering::greater);
+    test_cmp(big_decimal(max_scale - 3, huge_unscaled_number), big_decimal(max_scale, huge_unscaled_number), std::strong_ordering::greater);
+    test_cmp(big_decimal(max_scale, huge_unscaled_number), big_decimal(max_scale - 3, huge_unscaled_number), std::strong_ordering::less);
+    test_cmp(big_decimal(min_scale, huge_unscaled_number), big_decimal(min_scale + 3, huge_unscaled_number), std::strong_ordering::greater);
+    test_cmp(big_decimal(min_scale + 3, huge_unscaled_number), big_decimal(min_scale, huge_unscaled_number), std::strong_ordering::less);
+
+    // Test the comparison with numbers such that the difference is very close to the confidence window.
+    test_cmp(big_decimal("17.921310000000023424234"), big_decimal("1"), std::strong_ordering::greater); // just below
+    test_cmp(big_decimal("-18.891310000000023424234"), big_decimal("-1"), std::strong_ordering::less); // just above
+    test_cmp(big_decimal("-1"), big_decimal("-17.921310000000023424234"), std::strong_ordering::greater); // just below
+    test_cmp(big_decimal("1"), big_decimal("18.891310000000023424234"), std::strong_ordering::less); // just above
+
+    // Test fall-back to slow, should hit fall-back to slow and precise tri-compare.
+    test_cmp(big_decimal(-4, 1), big_decimal(-3, 15), std::strong_ordering::less);
+    test_cmp(big_decimal("1.1221310000000023424234"), big_decimal("1"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1"), big_decimal("1.1221310000000023424234"), std::strong_ordering::less);
+    test_cmp(big_decimal("1.88796E-4"), big_decimal("2.024E-4"), std::strong_ordering::less);
+    // The two numbers used in the following tests will generate very different scales, so the scale-compare
+    // algorithm will engage but at the end it should realize that the numbers are too close.
+    test_cmp(big_decimal("1.0000000000000000000001"), big_decimal("1.00000000000000000000010000000000000007"), std::strong_ordering::less);
+    test_cmp(big_decimal("1.00000000000000000000010000000000000007"), big_decimal("1.0000000000000000000001"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1.0000000000000000000001"), big_decimal("-1.00000000000000000000010000000000000007"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1.00000000000000000000010000000000000007"), big_decimal("-1.0000000000000000000001"), std::strong_ordering::less);
```

Comment:
You are missing here a test with very large scale but very close mantissa - for example
1e1000000 and 99e999998
This will convert both huge numbers to the same scale, but should be quick and won't suffer the original bug.

Reply:
Added a test in v5 (also with negative variant).